tests/test_client.py

```diff
@@ -108,5 +120,157 @@ async def launch_client() -> None:
+
+    class MockPahoClient(mqtt.Client):
+        def subscribe(
+            self,
+            topic: str
+            | tuple[str, mqtt.SubscribeOptions]
+            | list[tuple[str, mqtt.SubscribeOptions]]
+            | list[tuple[str, int]],
+            qos: int = 0,
+            options: mqtt.SubscribeOptions | None = None,
+            properties: mqtt.Properties | None = None,
+        ) -> tuple[int, int]:
+            assert client._outgoing_calls_sem is not None
+            assert client._outgoing_calls_sem.locked()
+            return super().subscribe(topic, qos, options, properties)
+
+        def unsubscribe(
+            self, topic: str | list[str], properties: mqtt.Properties | None = None
+        ) -> tuple[int, int]:
+            assert client._outgoing_calls_sem is not None
+            assert client._outgoing_calls_sem.locked()
+            return super().unsubscribe(topic, properties)
+
+        def publish(
+            self,
+            topic: str,
+            payload: PayloadType | None = None,
+            qos: int = 0,
+            retain: bool = False,
+            properties: mqtt.Properties | None = None,
+        ) -> mqtt.MQTTMessageInfo:
+            assert client._outgoing_calls_sem is not None
+            assert client._outgoing_calls_sem.locked()
+            return super().publish(topic, payload, qos, retain, properties)
+
+    monkeypatch.setattr(mqtt, "Client", MockPahoClient)
+
+    async with Client(HOSTNAME, max_concurrent_outgoing_calls=1) as client:
+        await client.subscribe(topic)
+        await client.unsubscribe(topic)
+        await client.publish(topic)
+
+
+async def test_client_websockets() -> None:
+    topic = TOPIC_HEADER + "websockets"
+
+    async def handle_messages(tg: anyio.abc.TaskGroup) -> None:
+        async with client.filtered_messages(topic) as messages:
+            async for message in messages:
+                assert message.topic == topic
+                tg.cancel_scope.cancel()
+
+    async with Client(
+        HOSTNAME,
+        8080,
+        transport="websockets",
+        websocket_path="/",
+        websocket_headers={"foo": "bar"},
+    ) as client:
+        async with anyio.create_task_group() as tg:
+            await client.subscribe(topic)
+            tg.start_soon(handle_messages, tg)
+            await client.publish(topic)
+
+
+async def test_client_pending_calls_threshold(caplog: pytest.LogCaptureFixture) -> None:
+    topic = TOPIC_HEADER + "pending_calls_threshold"
+
+    async with Client(HOSTNAME) as client:
+        nb_publish = client._pending_calls_threshold + 1
+
+        async with anyio.create_task_group() as tg:
+            for _ in range(nb_publish):
+                tg.start_soon(client.publish, topic)
+
+        assert caplog.record_tuples == [
+            (
+                "mqtt",
+                logging.WARNING,
+                f"There are {nb_publish} pending publish calls.",
```

Comment:
See my earlier comment about backpressure. The current "just log it and continue" approach of mine is *very* crude. 😅

In any case, good to have the behaviour covered in tests. 👍